.github/workflows/stack.yaml

```diff
@@ -0,0 +1,110 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest]
+        resolver: [nightly, lts-16, lts-14, lts-12, lts-11, lts-9, lts-6]
+        include:
+          - resolver: lts-6
+            ghc: 7.10.3
+            stack-yaml: stack-really-old.yaml
+          - resolver: lts-9
+            ghc: 8.0.2
+            stack-yaml: stack-really-old.yaml
+          - resolver: lts-11
+            ghc: 8.2.2
+            stack-yaml: stack-old.yaml
+          - resolver: lts-12
+            ghc: 8.4.4
+            stack-yaml: stack-old.yaml
+          - resolver: lts-14
+            stack-yaml: stack-coveralls.yaml
+          - resolver: lts-16
+            ghc: 8.8.4
+            stack-yaml: stack.yaml
+          # Latest stable for MacOS: ghc-8.8.4
+          - resolver: lts-16
+            os: macos-latest
+          # Latest stable for Windows: ghc-8.6.4
+          - resolver: lts-14
+            os: windows-latest
+            stack-yaml: stack-coveralls.yaml
+    env:
+      STACK_YAML: stack.yaml
+      STACK_ARGS: '--resolver ${{ matrix.resolver }}'
+      cache-version: v2 # bump up this version to invalidate currently stored cache
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Cache
+      id: cache
+      uses: actions/cache@v2
+      with:
+        path: |
+          ~/.stack
+          .stack-work
+        key: ${{ runner.os }}-${{ matrix.resolver }}-${{ env.cache-version }}
+        restore-keys: |
+          ${{ runner.os }}-${{ matrix.resolver }}-${{ env.cache-version }}
+    # Executable files somehow become incompatible after restoring on MacOS from a
+    # previous build, so it needs to be cleaned up. This is very inconvenient and will
+    # need to be fixed.
+    - name: MacOS workaround for failure due to setup-exe-cache
+      if: steps.cache.outputs.cache-hit == 'true' && matrix.os == 'macos-latest'
+      run: |
+        rm -r ~/.stack/setup-exe-cache
+        rm -r .stack-work
+    - name: Windows Cache
+      id: cache-windows
+      uses: actions/cache@v2
+      if: matrix.os == 'windows-latest'
+      with:
+        path: |
+          ~/AppData/Local/Programs/stack
+        key: ${{ runner.os }}-${{ matrix.resolver }}-programs-${{ env.cache-version }}
+        restore-keys: |
+          ${{ runner.os }}-${{ matrix.resolver }}-programs-${{ env.cache-version }}
+
+    # For ubuntu we can use fast ppa installation that also reduce cache size drastically
+    - name: Ubuntu install GHC
+      if: matrix.ghc != '' && matrix.os == 'ubuntu-latest'
+      run: |
+        sudo add-apt-repository ppa:hvr/ghc -y
+        sudo apt-get update
+        sudo apt-get install ghc-${{ matrix.ghc }}
+    - name: Install Stack and GHC
+      run: |
+        set -ex
+        curl -sSL https://get.haskellstack.org/ | sh -s - -f
+        curl -sSL https://gist.githubusercontent.com/lehins/fd36a8cc8bf853173437b17f6b6426ad/raw/4702d0252731ad8b21317375e917124c590819ce/git-modtime.hs -o git-modtime.hs
```

Comment:
I'd prefer to have this script inlined (maybe rewritten to Bash?). It feels uneasy to source it from a (potentially mutable) gist.

Reply:
I had it originally written in bash that worked on linux, but making it OS agnostic is a nightmare, so I decided instead to write in Haskell instead.

Here is a more detailed write up on this topic. I'll include the link to this answer in the CI yaml: https://stackoverflow.com/questions/60906336/stack-haskell-build-cache-of-source-files-with-github-actions/61178945#61178945

Source is immutable, because url contains the hash of the version, in other words the gist will probably change, but in order to get the updated version, the url will have to change.

Reply:
`defaults: { run: { shell: bash } }` should make `bash` the default shell even for Windows build. Are there compatibility challenges nevertheless?

Reply:
Yes, even for MacOS